src/main/java/uk/ac/sanger/sccp/stan/service/history/HistoryServiceImp.java

```diff
@@ -78,11 +80,131 @@ public History getHistoryForSampleId(int sampleId) {
         return getHistoryForSamples(samples);
     }
 
+    @Override
+    public History getHistory(String workNumber, String barcode, String externalName, String donorName){
+        //If Work number given, get history data from work
+        if(workNumber != null) {
+            return getHistoryForWorkNumberWithMatchingParams(workNumber, barcode, externalName, donorName);
+        }
+        List<Sample> samples = null;
+        //If barcode given,get all samples from the labware
+        if (barcode!=null) {
+            Labware lw = lwRepo.getByBarcode(barcode);
+            Set<Integer> tissueIds = lw.getSlots().stream()
+                    .flatMap(slot -> slot.getSamples().stream())
+                    .map(sam -> sam.getTissue().getId())
+                    .collect(toSet());
+             samples = sampleRepo.findAllByTissueIdIn(tissueIds);
+        }
+        List<Sample> filteredSamples = getFilteredSamples(externalName, donorName, samples);
+        return getHistoryForSamples(filteredSamples);
+    }
+
+    public List<Sample> getFilteredSamples(String externalName, String donorName, List<Sample> samples) {
+        List<Sample> retSamples = samples;
+        if (donorName != null) {
+            if(retSamples==null || retSamples.isEmpty()) {
```

Comment:
If retSamples is null, I think that means you haven't searched for samples yet; and if it's empty, it means you *have* already searched for samples, and you haven't found any. Those are different situations.